testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java

```diff
@@ -10537,6 +10537,29 @@ private static void checkArrayConcatAggFuncFails(SqlOperatorFixture t) {
         "[null, foo]", "CHAR(3) ARRAY NOT NULL");
     f2.checkScalar("array(null)",
         "[null]", "NULL ARRAY NOT NULL");
+    // check complex type
+    f2.checkScalar("array(row(1))", "[{1}]",
+        "RecordType(INTEGER NOT NULL EXPR$0) NOT NULL ARRAY NOT NULL");
+    f2.checkScalar("array(row(1, null))", "[{1, null}]",
+        "RecordType(INTEGER NOT NULL EXPR$0, NULL EXPR$1) NOT NULL ARRAY NOT NULL");
+    f2.checkScalar("array(row(null, 1))", "[{null, 1}]",
+        "RecordType(NULL EXPR$0, INTEGER NOT NULL EXPR$1) NOT NULL ARRAY NOT NULL");
+    f2.checkScalar("array(row(1, 2))", "[{1, 2}]",
+        "RecordType(INTEGER NOT NULL EXPR$0, INTEGER NOT NULL EXPR$1) NOT NULL ARRAY NOT NULL");
+    f2.checkFails("^array(row(1, 2), null)^",
+        "Parameters must be of the same type", false);
+    f2.checkFails("^array(null, row(1, 2))^",
```

Comment:
@mihaibudiu thanks for reviewing. the array use `SameOperandTypeChecker`, will call
https://github.com/apache/calcite/blob/56eb062ba2eae239e9fdda6891830cf2ec60b605/core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java#L1495

when element is null and row will return false. then checkFail. 

The reason is row type is a struct type. struct types require their fields are pairwise comparable in SqlTypeUtil.

Reply:
With the rule you give it is impossible to create a constant array with type `RecordType(...) ARRAY` which contains a `NULL` value. In other words, I think that the example that fails should actually be legal code. Clearly the NULL type is not isStruct(), but they can be both unified by a type which is a nullable struct.

Reply:
@mihaibudiu 
My original intention is just to test the case where the array contains null, such as
`array(1, null)`, `array('A', null)` ...
they can get expected result (can deduce correct type). however `array(row(1), null)` will cause error, because the element type is not same.
IMO, the end user may also use this form. Do you mean we don't need this `checkFail` ?

Reply:
To be clearly:
In fact, the error here is because of the effect of `type1.isStruct() != type2.isStruct()`, which works with `OperandTypeChecker`. I understand what you mean. You want `array(row(1, 2), null)` to return `RecordType(INTEGER EXPR $0, INTEGER EXPR$1) ARRAY NOT NULL`, row(1,2) & null will deduce leastRestrictive(or biggest) type row with nullable. In fact, it can indeed return this form in `derivedReturnType`. We only need to call derivedReturnType separately or comment out the line `type1.isStruct() != type2.isStruct()`.

But for the `SameOperandTypeChecker` of array, it needs to throw this error during checkOperandTypes. I understand that our checkFail can completely check the error of `OperandTypeChecker`. what do you think?

Reply:
Yes, you are right, I think that the test should pass with the output you have given.
This means that there is still one more bug left in the typechecking code.

Reply:
thanks mihai. you are right. the typechecking has a bug.
I have tested the calcite std array constructor and spark sql, it shows like below:

```
0: jdbc:calcite:model=src/test/resources/mode> select array[row(1,2), null];
+----------------+
|     EXPR$0     |
+----------------+
| [{1, 2}, null] |
+----------------+
```

```
spark-sql (default)> select array((1,2), null);
[{"col1":1,"col2":2},null]
```

It is indeed supported to return this form. I will open another issue to tracking/fixing this.

Reply:
If the fix is not too complicated it should probably be part of this PR.
Otherwise you have to check-in these tests which give the wrong result, and then modify them again. But perhaps it's reasonable to make a separate PR if it entails a lot of work.

Reply:
yes. I have updated the PR. In the additional commit, I have override the checkOperand methods. The logic of the overridden method is basically the same as the previous OperandTypeChecker, except that `SqlTypeUtil.isComparable` is replaced to `SqlTypeUtil.leastRestrictiveForComparison`. The first will reject row/struct type with NULL, the next will try to get `leastRestrictiveType`, the next is what we expect.

If you have time, PTAL again. thanks.